### DISCUSSION TITLE
MLFlow creating multiple model entries during multi-GPU training

### DISCUSSION THREAD

**Discussion by EvanZ** [mlflow/mlflow#7148] [General]
<img width="834" alt="Screen Shot 2022-10-24 at 11 05 51 AM" src="https://user-images.githubusercontent.com/1295082/197618460-3fe6cb2a-4c31-4850-a636-7a28c505fa49.png">
Not sure what is causing this but when I train my PyTorch models on a multi-GPU (4 GPU) EC2 instance, somehow 4 entries in MLFlow are being created. Is this a known issue or has someone seen this?

**Comment by JHibbard** [comment]
Hi EvanZ,

Could you share your training code? I can take a look and help troubleshoot.

Thanks!
Jim

**Comment by EvanZ** [comment]
Does this help?

```        
            trainer = pl.Trainer(
            fast_dev_run=args.fast,
            devices=args.gpus,
            accelerator=args.acc,
            callbacks=[early_stopping_cb,
                       checkpoint_cb,
                       TQDMProgressBar(refresh_rate=1)],
            precision=args.precision,
            strategy='ddp'
        )
        
        mlflow.set_tracking_uri('https://my.company.url')
        mlflow.pytorch.autolog(log_models=True)
        tags=create_tags(args.tags)
        
        with open('cli_args.json', 'w') as fp:
            json.dump(vars(args), fp)
        
        with mlflow.start_run(experiment_id=args.experiment, run_name=args.run_name) as run:
                        
            mlflow.set_tags(tags)
            trainer.fit(model, datamodule=dm)
            
            if args.log_model:
                mlflow.pytorch.log_model(model, args.path_log_model)
                
            mlflow.log_artifact('cli_args.json', 'cli/')
            mlflow.log_artifact(checkpoint_path, 'checkpoint/')
```

**Reply by JHibbard** [reply]
Thanks for the added info! Also, which version of MLflow are you on?

**Comment by EvanZ** [comment]
@JHibbard 1.30

**Comment by EvanZ** [comment]
This is really flumoxing me. I'm using a 8-GPU instance now and MLFlow is generating 8 model entires. 
<img width="661" alt="Screen Shot 2022-11-02 at 8 26 00 AM" src="https://user-images.githubusercontent.com/1295082/199530727-c1cfe251-a304-451b-866a-c4dc550b01d6.png">

**Comment by cao-nv** [comment]
Are you using distributed training? 
In that case, you should on do the logging only on the RANK 0 GPU
somethings like: 
```python
LOCAL_RANK = os.environ["LOCAL_RANK"] # or any way to get local rank 
if LOCAL_RANK == 0: 
    mlflow.start_run(...)
```

**Reply by EvanZ** [reply]
I am not using distributed training. This is just one multi-GPU instance.

**Reply by cao-nv** [reply]
A new experiment is created when you call `mlflow.start_run(...)`, please check if you are calling it in a loop.
